### DISCUSSION TITLE
How to show the result of offline CI for SYCL backend

### DISCUSSION THREAD

**Discussion by NeoZhangJianyu** [ggml-org/llama.cpp#6399] [General]
Background:
Current CI of SYCL backend is only building code, without running unit test.
Run the unit test need Intel GPU, which is not provided by github runner.
It depends on developer running local unit test before create PR. 
For many reasons, the unit test pass rate is easy to be broken from 100%.
That leads to the result of OPs have obvious error, performance drop.

That impacts the whole SYCL backend quality and user experience.

Solution:
   Offline CI for SYCL backend:
    1. github doesn't recommend to setup self-hosted runner for public repo, due to security issue.
    2. I run the unit test/performance for SYCL backend in my personal computer with Intel iGPU. 
        I choose the existed commits which change SYCL backend.
        I could know the quality and performance of every commit.

Question:
  How to show the result of offline CI for SYCL backend？
  I hope to publish the result of every key commit.
  It will help developer to improve the quality and user to choose the release with best quality and performance.

Alternative:
a. Update the summary of unit test result of SYCL backend in README-sycl.md
b. Insert the unit test pass rate in the title of merged PR.
c. Update the summary of test result in another public repo MD file and insert the link in README-sycl.md. So user is easy to read it to check the latest SYCL backend test result.

I choose c.
How about your idea?

@ggerganov @slaren @airMeng

**Comment by phymbert** [comment]
We successfully deployed an approach with self-hosted github runner for CUDA, maybe the same can be done for SYCL ?
- https://github.com/ggml-org/ci/pull/2
- #6283
- #6233

**Reply by NeoZhangJianyu** [reply]
Thank your sharing.
But github doesn't recommend to use self-hosted runner for public repo.
It has security issue: it would run the attack code on the self-hosted runner.
It's ok if the self-hosted runner is on public cloud, due to there is powerful security protect.
But I will run in personal computer, I can't cover the risk.

**Reply by ggerganov** [reply]
Do you know if there is such hardware in Azure Cloud by any chance? If so, I can rent a dedicated node as part of `ggml-ci`

**Reply by phymbert** [reply]
We took this into consideration:
- the runner is JIT for one job
- the workflow is running inside a rootless docker container
- the workflow cannot be modified by the PR

**Reply by ggerganov** [reply]
@phymbert The problem with this option is that I cannot share the admin token with them

**Reply by phymbert** [reply]
I was anticipating running it on an Azure VM on ggml the same way it works for CUDA. Dumb question will it not work with [Dv3 series](https://learn.microsoft.com/en-us/azure/virtual-machines/dv3-dsv3-series) ?

**Reply by NeoZhangJianyu** [reply]
Yes, I will check public cloud. Thank you!

**Reply by NeoZhangJianyu** [reply]
@ggerganov  Looks like on Intel Dev Cloud provide such VM:

https://console.cloud.intel.com/compute/reserve.

Xeon with Intel Flex GPU (16GB)

<img width="592" alt="intel_dev_cloud_flex" src="https://github.com/ggerganov/llama.cpp/assets/46982523/edc00da0-52d1-4822-9227-061953bd0d95">

**Reply by ggerganov** [reply]
The AI Grant sponsorship that I'm using for `ggml-ci` only provides credits for Azure Cloud, so I won't be able to rent the VM on Intel Dev Cloud. If you can rent a machine, then it should not be a problem to add it to the CI

**Reply by NeoZhangJianyu** [reply]
Sorry, I'm an individual developer. I can only contribute code in my spare time. :)
Anyway, it's an alternative.

**Reply by phymbert** [reply]
sorry I did not get the answer @NeoZhangJianyu @ggerganov : will not it work with Azure [Dv3 series](https://learn.microsoft.com/en-us/azure/virtual-machines/dv3-dsv3-series) ?

**Reply by NeoZhangJianyu** [reply]
SYCL backend support Intel iGPU/dGPU by now.
Dv3 series includes Intel Xeon CPUs. This node has no Intel iGPU/dGPU looks like.
If I'm wrong, please correct me.

**Reply by tonym97** [reply]
Just to add here, I can look into options on Intel Tiber cloud if it would be a reasonable workflow for you @ggerganov

**Reply by ggerganov** [reply]
It doesn't matter in which cloud the machine is - as long as we have SSH access to it, we can setup `ggml-ci` to run on it

**Reply by tonym97** [reply]
Perfect.

**Reply by tonym97** [reply]
How many systems would you need @ggerganov with Intel GPUs?  What's your typical CI/CD requirements?

**Reply by ggerganov** [reply]
One system should be enough so let's start with that. In the future if you have a variety of Intel GPU hardware that you would like to be exercised we can add more nodes

The CI requires at least a 128 GB disk since it downloads a few models that can be tens of GBs. It runs Ubuntu 22.04 and requires the following packages (+ any SYCL related stuff) to be installed:

```bash
sudo apt install vim curl cmake g++ python3-pip python3-venv unzip ccache
```

No other requirements are necessary except for an SSH access so I can login and setup [ggml-ci](https://github.com/ggml-org/ci) to run

**Reply by tonym97** [reply]
Got it. Will get back to you in short order. Thanks!

**Reply by tonym97** [reply]
@ggerganov I have some buy in, I wanted to chat with you to align hardware requirements since you already provided software above.  I'm going to drop you an email if that's ok.
